File: src/caco/cli/cache.py
"""Cache management commands: cache list/clear/clean."""

import sys
from pathlib import Path

import click
from rich.table import Table

from caco import db
from caco.config import get_cache_dir

from caco.cli import (
    cli,
    console,
    err_console,
    resolve_wad_query,
)


# =============================================================================
# Internal helpers
# =============================================================================


def _format_size(size_bytes: int) -> str:
    """Format bytes as human-readable size."""
    for unit in ["B", "KB", "MB", "GB"]:
        if size_bytes < 1024:
            if unit == "B":
                return f"{size_bytes} {unit}"
            return f"{size_bytes:.1f} {unit}"
        size_bytes /= 1024
    return f"{size_bytes:.1f} TB"


def _list_orphaned_files(cache_dir: Path, plain: bool) -> None:
    """List orphaned files in the cache directory."""
    if not cache_dir.exists():
        if plain:
            click.echo("No orphaned files")
        else:
            console.print("[dim]Cache directory does not exist[/dim]")
        return

    orphans = []
    total_size = 0

    for path in cache_dir.iterdir():
        if path.is_file():
            # Check if any WAD references this file
            wad = db.get_wad_by_cached_filename(path.name)
            if not wad:
                size = path.stat().st_size
                orphans.append((path, size))
                total_size += size

    if plain:
        click.echo("Filename\tSize")
        for path, size in orphans:
            click.echo(f"{path.name}\t{size}")
        click.echo(f"\nTotal: {total_size}")
    else:
        if not orphans:
            console.print("[dim]No orphaned files found[/dim]")
            return

        table = Table(title=f"Orphaned Files ({len(orphans)} files)")
        table.add_column("Filename")
        table.add_column("Size", justify="right")

        for path, size in orphans:
            table.add_row(path.name, _format_size(size))

        console.print(table)
        console.print(f"\n[bold]Total:[/bold] {_format_size(total_size)}")
        console.print(f"[dim]Use 'caco cache clean' to remove orphaned files[/dim]")


def _clear_all_cache(cache_dir: Path, dry_run: bool, yes: bool) -> None:
    """Clear the entire cache."""
    cached_wads = db.get_cached_wads()
    total_size = 0
    files_to_delete = []

    for wad in cached_wads:
        # Only delete idgames sources - they can always be re-downloaded
        # Local files are the user's originals, URLs may not be re-downloadable
        if wad.get("source_type") != "idgames":
            continue

        path = Path(wad["cached_path"])
        if path.exists():
            size = path.stat().st_size
            total_size += size
            files_to_delete.append((wad, path, size))

    if not files_to_delete:
        console.print("[dim]Cache is empty (no idgames WADs cached)[/dim]")
        return

    # Preview
    console.print(f"\n[bold]Files to delete ({len(files_to_delete)}):[/bold]\n")
    for wad, path, size in files_to_delete[:10]:
        console.print(f"  [dim][{wad['id']}][/dim] {wad['title']} ({_format_size(size)})")
    if len(files_to_delete) > 10:
        console.print(f"  [dim]... and {len(files_to_delete) - 10} more[/dim]")

    console.print(f"\n[bold]Total:[/bold] {_format_size(total_size)}")

    if dry_run:
        console.print("\n[dim]No changes made (dry run)[/dim]")
        return

    if not yes:
        if not click.confirm("\nDelete all cached files?"):
            console.print("[dim]Cancelled[/dim]")
            return

    # Delete files and update database
    deleted = 0
    freed = 0
    for wad, path, size in files_to_delete:
        try:
            path.unlink()
            freed += size
            deleted += 1
            db.clear_cached_path(wad["id"])
        except OSError as e:
            err_console.print(f"[red]Failed to delete {path}: {e}[/red]")

    console.print(f"\n[green]Deleted {deleted} file(s), freed {_format_size(freed)}[/green]")


def _clear_specific_cache(query: str, cache_dir: Path, dry_run: bool, yes: bool) -> None:
    """Clear cache for specific WADs."""
    wads = resolve_wad_query(query, mode="multiple", yes=yes)
    if not wads:
        return

    # Filter to only cached WADs (skip local source and non-cached)
    files_to_delete = []
    total_size = 0

    for wad in wads:
        # Only delete idgames sources - they can always be re-downloaded
        if wad.get("source_type") != "idgames":
            continue
        if not wad.get("cached_path"):
            continue
        path = Path(wad["cached_path"])
        if path.exists():
            size = path.stat().st_size
            total_size += size
            files_to_delete.append((wad, path, size))

    if not files_to_delete:
        console.print("[dim]Selected WAD(s) are not cached or not from idgames[/dim]")
        return

    # Preview
    console.print(f"\n[bold]Files to delete ({len(files_to_delete)}):[/bold]\n")
    for wad, path, size in files_to_delete[:10]:
        console.print(f"  [dim][{wad['id']}][/dim] {wad['title']} ({_format_size(size)})")
    if len(files_to_delete) > 10:
        console.print(f"  [dim]... and {len(files_to_delete) - 10} more[/dim]")

    console.print(f"\n[bold]Total:[/bold] {_format_size(total_size)}")

    if dry_run:
        console.print("\n[dim]No changes made (dry run)[/dim]")
        return

    # Delete
    deleted = 0
    freed = 0
    for wad, path, size in files_to_delete:
        try:
            path.unlink()
            freed += size
            deleted += 1
            db.clear_cached_path(wad["id"])
        except OSError as e:
            err_console.print(f"[red]Failed to delete {path}: {e}[/red]")

    console.print(f"\n[green]Deleted {deleted} file(s), freed {_format_size(freed)}[/green]")


# =============================================================================
# Cache group
# =============================================================================


@cli.group(name="cache")
def cache_cmd():
    """Manage WAD cache."""
    pass


@cache_cmd.command(name="list")
@click.option("--plain", is_flag=True, help="Output as TSV (for scripting)")
@click.option("--orphans", is_flag=True, help="Show orphaned files (not in database)")
def cache_list(plain: bool, orphans: bool):
    """List cached WAD files and total cache size."""
    cache_dir = get_cache_dir()

    if orphans:
        _list_orphaned_files(cache_dir, plain)
        return

    cached_wads = db.get_cached_wads()
    total_size = 0
    entries = []

    for wad in cached_wads:
        path = Path(wad["cached_path"])
        if path.exists():
            size = path.stat().st_size
            total_size += size
            entries.append({
                "id": wad["id"],
                "title": wad["title"],
                "filename": path.name,
                "size": size,
                "path": str(path),
            })

    if plain:
        click.echo("ID\tTitle\tFilename\tSize")
        for e in entries:
            click.echo(f"{e['id']}\t{e['title']}\t{e['filename']}\t{e['size']}")
        click.echo(f"\nTotal: {total_size}")
    else:
        if not entries:
            console.print("[dim]Cache is empty[/dim]")
            console.print(f"[dim]Cache location: {cache_dir}[/dim]")
            return

        table = Table(title=f"Cached WADs ({len(entries)} files)")
        table.add_column("ID", style="dim")
        table.add_column("Title", style="cyan")
        table.add_column("Filename")
        table.add_column("Size", justify="right")

        for e in entries:
            table.add_row(
                str(e["id"]),
                e["title"],
                e["filename"],
                _format_size(e["size"]),
            )

        console.print(table)
        console.print(f"\n[bold]Total cache size:[/bold] {_format_size(total_size)}")
        console.print(f"[dim]Cache location: {cache_dir}[/dim]")


@cache_cmd.command(name="clear")
@click.argument("query", required=False)
@click.option("--all", "clear_all", is_flag=True, help="Clear entire cache")
@click.option("--dry-run", is_flag=True, help="Show what would be deleted")
@click.option("--yes", "-y", is_flag=True, help="Skip confirmation prompt")
def cache_clear(query: str | None, clear_all: bool, dry_run: bool, yes: bool):
    """Remove cached WAD files.

    Without arguments, shows help. Use --all to clear entire cache,
    or specify QUERY to clear specific WADs.

    \b
    Examples:
        caco cache clear --all           # Clear entire cache
        caco cache clear 1,3,5           # Clear specific WAD IDs
        caco cache clear status:finished # Clear all finished WADs
    """
    if not query and not clear_all:
        click.echo(click.get_current_context().get_help())
        return

    cache_dir = get_cache_dir()

    if clear_all:
        _clear_all_cache(cache_dir, dry_run, yes)
    else:
        _clear_specific_cache(query, cache_dir, dry_run, yes)


@cache_cmd.command(name="clean")
@click.option("--dry-run", is_flag=True, help="Show what would be deleted")
@click.option("--yes", "-y", is_flag=True, help="Skip confirmation prompt")
def cache_clean(dry_run: bool, yes: bool):
    """Remove orphaned files from cache.

    Orphaned files are files in the cache directory that are not
    tracked by any WAD in the database.
    """
    cache_dir = get_cache_dir()
    if not cache_dir.exists():
        console.print("[dim]Cache directory does not exist[/dim]")
        return

    # Find all orphaned files
    orphans = []
    total_size = 0

    for path in cache_dir.iterdir():
        if path.is_file():
            wad = db.get_wad_by_cached_filename(path.name)
            if not wad:
                size = path.stat().st_size
                orphans.append((path, size))
                total_size += size

    if not orphans:
        console.print("[dim]No orphaned files found[/dim]")
        return

    # Preview
    console.print(f"\n[bold]Orphaned files ({len(orphans)}):[/bold]\n")
    for path, size in orphans[:10]:
        console.print(f"  {path.name} ({_format_size(size)})")
    if len(orphans) > 10:
        console.print(f"  [dim]... and {len(orphans) - 10} more[/dim]")

    console.print(f"\n[bold]Total:[/bold] {_format_size(total_size)}")

    if dry_run:
        console.print("\n[dim]No changes made (dry run)[/dim]")
        return

    if not yes:
        if not click.confirm("\nDelete orphaned files?"):
            console.print("[dim]Cancelled[/dim]")
            return

    # Delete
    deleted = 0
    freed = 0
    for path, size in orphans:
        try:
            path.unlink()
            freed += size
            deleted += 1
        except OSError as e:
            err_console.print(f"[red]Failed to delete {path}: {e}[/red]")

    console.print(f"\n[green]Deleted {deleted} orphaned file(s), freed {_format_size(freed)}[/green]")

File: crates/caco-cli/src/commands/cache.rs
//! `caco cache` — list, clear, and prune cached WAD files.

use clap::Subcommand;
use rusqlite::Connection;

use caco_core::config;
use caco_core::db;
use caco_core::utils::format_size;

use crate::output::OutputFormat;
use crate::resolve;

#[derive(Subcommand)]
pub enum CacheCommand {
    /// List cached WADs
    List {
        /// Output format: plain | json | table
        #[arg(short = 'o', long = "output", default_value = "table")]
        output: String,
        /// Show orphaned cache files
        #[arg(long)]
        orphans: bool,
    },
    /// Remove cached WAD files
    Clear {
        /// WAD query
        query: Vec<String>,
        /// Clear all cached files
        #[arg(long)]
        all: bool,
        /// Preview changes
        #[arg(long)]
        dry_run: bool,
        /// Skip confirmation
        #[arg(short = 'y', long)]
        yes: bool,
    },
    /// Remove orphaned cache files
    Prune {
        /// Preview changes
        #[arg(long)]
        dry_run: bool,
        /// Skip confirmation
        #[arg(short = 'y', long)]
        yes: bool,
    },
}

pub fn run(conn: &Connection, cmd: &CacheCommand) -> Result<(), String> {
    match cmd {
        CacheCommand::List { output, orphans } => {
            let format: OutputFormat = output.parse()?;
            if *orphans {
                list_orphans(conn, format)
            } else {
                list_cached(conn, format)
            }
        }
        CacheCommand::Clear {
            query,
            all,
            dry_run,
            yes,
        } => {
            if *all {
                clear_all(conn, *dry_run, *yes)
            } else {
                clear_query(conn, query, *dry_run, *yes)
            }
        }
        CacheCommand::Prune { dry_run, yes } => prune(conn, *dry_run, *yes),
    }
}

fn list_cached(conn: &Connection, format: OutputFormat) -> Result<(), String> {
    let wads = db::get_cached_wads(conn).map_err(|e| e.to_string())?;
    if wads.is_empty() && format != OutputFormat::Json {
        println!("No cached WADs.");
        return Ok(());
    }

    // Pre-compute per-row size once.
    let rows: Vec<(i64, &str, &str, Option<&str>, u64)> = wads
        .iter()
        .map(|w| {
            let size = w
                .cached_path
                .as_deref()
                .and_then(|p| std::fs::metadata(p).ok())
                .map(|m| m.len())
                .unwrap_or(0);
            (
                w.id,
                w.title.as_str(),
                w.filename.as_deref().unwrap_or(""),
                w.cached_path.as_deref(),
                size,
            )
        })
        .collect();
    let total_size: u64 = rows.iter().map(|r| r.4).sum();

    match format {
        OutputFormat::Table => {
            use comfy_table::{Cell, CellAlignment, Table, presets};
            let mut table = Table::new();
            table
                .load_preset(presets::UTF8_FULL_CONDENSED)
                .set_header(vec!["ID", "Title", "Filename", "Size"]);
            for (id, title, filename, _path, size) in &rows {
                table.add_row(vec![
                    Cell::new(id).set_alignment(CellAlignment::Right),
                    Cell::new(title),
                    Cell::new(filename),
                    Cell::new(format_size(*size)).set_alignment(CellAlignment::Right),
                ]);
            }
            println!("{table}");
            println!(
                "{} cached WAD(s), {} total",
                rows.len(),
                format_size(total_size)
            );
        }
        OutputFormat::Plain => {
            println!("ID\tTitle\tFilename\tSize");
            for (id, title, filename, _path, size) in &rows {
                println!("{id}\t{title}\t{filename}\t{}", format_size(*size));
            }
            println!("total\t{}\t{}", rows.len(), format_size(total_size));
        }
        OutputFormat::Json => {
            let items: Vec<_> = rows
                .iter()
                .map(|(id, title, filename, path, size)| {
                    serde_json::json!({
                        "id": id,
                        "title": title,
                        "filename": filename,
                        "cached_path": path,
                        "size": size,
                    })
                })
                .collect();
            let out = serde_json::json!({
                "count": rows.len(),
                "total_size": total_size,
                "items": items,
            });
            println!("{}", serde_json::to_string_pretty(&out).unwrap_or_default());
        }
    }
    Ok(())
}

fn list_orphans(conn: &Connection, format: OutputFormat) -> Result<(), String> {
    let cache_dir = config::get_cache_dir();
    if !cache_dir.is_dir() && format != OutputFormat::Json {
        println!("No cache directory found.");
        return Ok(());
    }

    let mut orphans = Vec::new();
    if cache_dir.is_dir()
        && let Ok(entries) = std::fs::read_dir(&cache_dir)
    {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                let filename = path
                    .file_name()
                    .and_then(|n| n.to_str())
                    .unwrap_or("")
                    .to_string();
                if db::get_wad_by_cached_filename(conn, &filename)
                    .map_err(|e| e.to_string())?
                    .is_none()
                {
                    let size = std::fs::metadata(&path).map(|m| m.len()).unwrap_or(0);
                    orphans.push((path, filename, size));
                }
            }
        }
    }

    if orphans.is_empty() && format != OutputFormat::Json {
        println!("No orphaned cache files.");
        return Ok(());
    }

    match format {
        OutputFormat::Table => {
            println!("{} orphaned file(s):", orphans.len());
            for (_, name, size) in &orphans {
                println!("  {name} ({})", format_size(*size));
            }
        }
        OutputFormat::Plain => {
            println!("Filename\tSize");
            for (_, name, size) in &orphans {
                println!("{name}\t{}", format_size(*size));
            }
        }
        OutputFormat::Json => {
            let items: Vec<_> = orphans
                .iter()
                .map(|(path, name, size)| {
                    serde_json::json!({
                        "filename": name,
                        "path": path.to_string_lossy(),
                        "size": size,
                    })
                })
                .collect();
            let out = serde_json::json!({
                "count": orphans.len(),
                "items": items,
            });
            println!("{}", serde_json::to_string_pretty(&out).unwrap_or_default());
        }
    }
    Ok(())
}

fn clear_all(conn: &Connection, dry_run: bool, yes: bool) -> Result<(), String> {
    let wads = db::get_cached_wads(conn).map_err(|e| e.to_string())?;
    // Only clear idgames-sourced WADs (always re-downloadable)
    let clearable: Vec<_> = wads
        .iter()
        .filter(|w| w.source_type == db::SourceType::Idgames)
        .collect();

    if clearable.is_empty() {
        println!("No clearable cached files (only idgames sources can be cleared).");
        return Ok(());
    }

    if dry_run {
        println!("Would clear {} cached file(s).", clearable.len());
        return Ok(());
    }

    if !yes && !resolve::confirm(&format!("Clear {} cached file(s)?", clearable.len())) {
        return Err("Aborted.".to_string());
    }

    let mut cleared = 0;
    for wad in &clearable {
        if let Some(ref path) = wad.cached_path {
            let _ = std::fs::remove_file(path);
        }
        db::clear_cached_path(conn, wad.id).map_err(|e| e.to_string())?;
        cleared += 1;
    }
    println!("Cleared {cleared} cached file(s).");
    Ok(())
}

fn clear_query(
    conn: &Connection,
    query: &[String],
    dry_run: bool,
    yes: bool,
) -> Result<(), String> {
    if query.is_empty() {
        return Err("No query specified. Use --all to clear all.".to_string());
    }
    let wads = crate::resolve::resolve_wads(
        conn,
        query,
        crate::resolve::ResolveMode::Multiple,
        yes,
        false,
    )?;

    let clearable: Vec<_> = wads.iter().filter(|w| w.cached_path.is_some()).collect();
    if clearable.is_empty() {
        println!("No cached files for the specified WAD(s).");
        return Ok(());
    }

    if dry_run {
        println!("Would clear {} cached file(s).", clearable.len());
        return Ok(());
    }

    let mut cleared = 0;
    for wad in &clearable {
        if let Some(ref path) = wad.cached_path {
            let _ = std::fs::remove_file(path);
        }
        db::clear_cached_path(conn, wad.id).map_err(|e| e.to_string())?;
        cleared += 1;
    }
    println!("Cleared {cleared} cached file(s).");
    Ok(())
}

fn prune(conn: &Connection, dry_run: bool, yes: bool) -> Result<(), String> {
    let cache_dir = config::get_cache_dir();
    if !cache_dir.is_dir() {
        println!("No cache directory found.");
        return Ok(());
    }

    let mut orphans = Vec::new();
    if let Ok(entries) = std::fs::read_dir(&cache_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                let filename = path.file_name().and_then(|n| n.to_str()).unwrap_or("");
                if db::get_wad_by_cached_filename(conn, filename)
                    .map_err(|e| e.to_string())?
                    .is_none()
                {
                    orphans.push(path);
                }
            }
        }
    }

    if orphans.is_empty() {
        println!("No orphaned cache files to prune.");
        return Ok(());
    }

    if dry_run {
        println!("Would prune {} orphaned file(s).", orphans.len());
        return Ok(());
    }

    if !yes && !resolve::confirm(&format!("Prune {} orphaned file(s)?", orphans.len())) {
        return Err("Aborted.".to_string());
    }

    let mut pruned = 0;
    for path in &orphans {
        if std::fs::remove_file(path).is_ok() {
            pruned += 1;
        }
    }
    println!("Pruned {pruned} orphaned file(s).");
    Ok(())
}
